Firestore/core/src/util/executor_libdispatch.h

```diff
@@ -72,28 +57,36 @@ class ExecutorLibdispatch : public Executor {
   void Execute(Operation&& operation) override;
   void ExecuteBlocking(Operation&& operation) override;
   DelayedOperation Schedule(Milliseconds delay,
-                            TaggedOperation&& operation) override;
+                            Tag tag,
+                            Operation&& operation) override;
 
-  void RemoveFromSchedule(TimeSlotId to_remove);
-
-  bool IsScheduled(Tag tag) const override;
-  absl::optional<TaggedOperation> PopFromSchedule() override;
+  bool IsTagScheduled(Tag tag) const override;
+  bool IsIdScheduled(Id id) const override;
+  Task* PopFromSchedule() override;
 
   dispatch_queue_t dispatch_queue() const {
     return dispatch_queue_;
   }
 
  private:
-  using ScheduleMap = std::unordered_map<TimeSlotId, TimeSlot*>;
+  using ScheduleMap = std::unordered_map<Id, Task*>;
   using ScheduleEntry = ScheduleMap::value_type;
 
-  TimeSlotId NextId();
+  void Complete(Task* task) override;
+  void Cancel(Id operation_id) override;
+
+  static void InvokeAsync(void* raw_task);
+  static void InvokeSync(void* raw_task);
+
+  Id NextIdLocked();
+
+  mutable std::mutex mutex_;
 
   dispatch_queue_t dispatch_queue_;
-  // Stores non-owned pointers to `TimeSlot`s.
-  // Invariant: if a `TimeSlot` is in `schedule_`, it's a valid pointer.
+
+  std::unordered_set<Task*> async_tasks_;
```

Comment:
`async_tasks_` (which I've renamed to `tasks_` because synchronous tasks are present too) tracks all tasks.

`schedule_` is an index of tasks by their Id, and only contains scheduled tasks. Cancellation of scheduled tasks is common (e.g. the idle timer task is canceled on every write), and forcing this to be a linear search of everything in the async queue is more expensive than I want it to be.

A single bimap would be sufficient for both usages, though it would require non-scheduled tasks to also have an Id.

I suspect we could get rid of the schedule altogether if `DelayedOperation` retained a pointer to the backing task, but trying to implement that is something I want to defer.

I've added comments describing the separation.

Reply:
That's a good point. Do you think it's worth appending this to the existing comment? Rough sketch: `(while under normal operation the number of tasks on the schedule is small and doesn't grow much, unusual usage of the SDK could result in thousands of tasks being added)`.

Reply:
I think it's fairly reasonable (and our usual practice) to just pick the structures that give the right asymptotic complexity. It's only when intentionally choosing a structure that isn't intuitively right that it's worth commenting on why that's the right choice. In this case, if we only had the `tasks_` set I'd expect a comment on why the linear search is reasonable.